### ISSUE TITLE
.

### ISSUE THREAD

**Issue by kdmccormick** [overhangio/2u-tutor-adoption#10]

**Comment by regisb** [comment]
I believe that working on this matter will help us resolve #32: a robust and comprehensive plugin mechanism will be necessary to extract the LMS and CMS from the core.

**Comment by kdmccormick** [comment]
From #5:
> Well, when we move the LMS and the CMS out of the core, I expect that the [current tutor documentation](https://docs.tutor.overhang.io/) is still very much going to remain the same. It's actually the "core" that will be moved away from tutor and to a different project, with its own documentation. So if we end up with two new documentation pages where we have an exhaustive list of patches and settings, then this list will remain there.

Interesting!

If we start with this world:
```
tutor:
config, patches, templates, plugins, lms, cms
----------------------------------------------
^            ^ 
|            |
tutor-forum  tutor-discovery   etc...
```

I had imagined (assumed?) that the LMS/CMS extraction would yield a world like:

```
tutor:
config, patches, templates, plugins
---------------------------------------
^          ^          ^            ^ 
|          |          |            |
tutor-lms  tutor-cms  tutor-forum  tutor-discovery     ...etc
```
ie, *Tutor is a generic tool for configuring and running pluggable webservices, with Open edX LMS & CMS being its flagship plugins.*

But it sounds like you envision something more like:

```
tutorcore:
config, patches, templates, plugins
-----------------------------------
^        ^
|        |
|        other-tutorcore-based-tools?
|
|
tutor: lms, cms
--------------------------------------
^             ^ 
|             |
tutor-forum   tutor-discovery   etc...
```

ie, *Tutor-Core is a generic tool for configuring and running pluggable webservices, and Tutor (which runs Open edX LMS & CMS) is its flagship plugin, which itself can be customized through Tutor-specific Tutor-Core plugins.*

Do I have that right?

**Comment by regisb** [comment]
> Tutor-Core is a generic tool for configuring and running pluggable webservices, and Tutor (which runs Open edX LMS & CMS) is its flagship plugin, which itself can be customized through Tutor-specific Tutor-Core plugins.

Yes, this is exactly right :) As a devops I spend a lot of time installing and managing software, and I'm always amazed how frequently how different projects need to reinvent the same thing to make their open source software installable. For instance:

- Discourse has [discourse_docker](https://github.com/discourse/discourse_docker)
- Docker-Mailserver (_excellent_ project) has [setup.sh](https://github.com/docker-mailserver/docker-mailserver/blob/master/setup.sh).
- Sentry has [install.sh](https://github.com/getsentry/self-hosted/blob/master/install.sh)

There are tons of tools out there that would benefit from a simple configuration/environment/cli management interface. I felt really bummed and surprised that I didn't discover an existing, more appropriate solution when I started working on Tutor. The closest I could find was Ansible but there were several issues with that:

- Logging is terrible
- CLI is non-intuitive
- Too damn slow
